### ISSUE TITLE
How to change the IoU threshold when calculating mAP.

### ISSUE THREAD

**Issue by ue17** [ultralytics/ultralytics#4577]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

In the segmentation model, during validation, the mAP is calculated with IoU threshold = 0.5.
I would like to change the comparison with the validation data e.g. IoU threshold = 0.4 or 0.3.
Nothing is stated in the docs. How can I validate the above?

**Comment by glenn-jocher** [comment]
@ue17 hello,

Thank you for reaching out!

To adjust the IoU threshold when calculating mAP in YOLOv8, it's feasible to modify the "iou_threshold" parameter in the evaluation function or script responsible for the calculation. "iou_threshold" is commonly set to 0.5 by default, but it can be manually adjusted according to your preferences, like 0.4 or 0.3 you mentioned.

However, it's important to point out that changes in this threshold will affect model performance evaluations. Lowering the IoU threshold might increase mAP, as predictions need not be as precise to be considered a match with the ground truths. Conversely, raising the IoU threshold might decrease mAP, as predictions need to be more accurate.

Normally, this parameter should be adjusted depending on your specific application requirements and interpretation of model accuracy.

I hope this information helps! If there are any more questions regarding this or any other topic, kindly feel free to ask.

Best regards,
Glenn Jocher

**Comment by ue17** [comment]
@glenn-jocher 
Thank you.
Can you please tell me how to write a specific script?

**Comment by glenn-jocher** [comment]
@ue17, 

To write a script for adjusting the IoU threshold, you would first need to locate the evaluation function or script in the YOLOv8's source code where the mAP is calculated based on the 'iou_threshold' parameter. This can typically be found in the test or evaluation files. 

Once you've identified where the IoU threshold is set (often `iou_threshold=0.5`), you should simply need to change this value to your preferred threshold.

Remember that being comfortable with Python and understanding how the Ultralytics YOLOv8 repository is written are important prerequisites to making these adjustments. 

I hope this helps guide your next steps. Let me know if you have any other concerns, I'll be glad to assist further.

Best Regards,
Glenn Jocher

**Comment by ue17** [comment]
Thanks.

I have checked the code and believe the following source is applicable.
"engine/validator.py"

----
    def match_predictions(self, pred_classes, true_classes, iou):
        """
        Matches predictions to ground truth objects (pred_classes, true_classes) using IoU.

        Args:
            pred_classes (torch.Tensor): Predicted class indices of shape(N,).
            true_classes (torch.Tensor): Target class indices of shape(M,).
            iou (torch.Tensor): An NxM tensor containing the pairwise IoU values for predictions and ground of truth

        Returns:
            (torch.Tensor): Correct tensor of shape(N,10) for 10 IoU thresholds.
        """
        correct = np.zeros((pred_classes.shape[0], self.iouv.shape[0])).astype(bool)
        correct_class = true_classes[:, None] == pred_classes
        #for i, iouv in enumerate(self.iouv):
        for i, iouv in enumerate([0.4]):
            x = torch.nonzero(iou.ge(iouv) & correct_class)  # IoU > threshold and classes match
            if x.shape[0]:
                # Concatenate [label, detect, iou]
                matches = torch.cat((x, iou[x[:, 0], x[:, 1]].unsqueeze(1)), 1).cpu().numpy()
                if x.shape[0] > 1:
                    matches = matches[matches[:, 2].argsort()[::-1]]
                    matches = matches[np.unique(matches[:, 1], return_index=True)[1]]
                    # matches = matches[matches[:, 2].argsort()[::-1]]
                    matches = matches[np.unique(matches[:, 0], return_index=True)[1]]
                correct[matches[:, 1].astype(int), i] = True
        return torch.tensor(correct, dtype=torch.bool, device=pred_classes.device)
---

I think I need to change the iouv of this source, but I could not get it to work.
Could you please be more specific?

**Comment by glenn-jocher** [comment]
@ue17 hello,

In the code provided, it seems the IoU threshold values are stored in the `iou` variable and then passed to the function `match_predictions`. Modifying this value here might allow you to change the IoU threshold.

You can see that the function `match_predictions` loops over a set of IoU threshold values specified by the `BaseValidator.iouv` attribute when comparing predicted and actual classes for each prediction. In your example, you've replaced self.iouv with [0.4], and as a result the evaluation is performed with an IoU threshold of 0.4.

For the `BaseValidator.iouv` part, it is initialized elsewhere in the code, most likely in the constructor of the `BaseValidator` where all member attributes get their initial values. 

You should be able to modify the IoU thresholds by finding where these `iou` values are originally set and adjusting them accordingly.

Remember, when you set the `iou` threshold, make sure it is in the correct format (a list of thresholds to iterate through). Review the original list values (usually range from 0.50 to 0.95, incremented by 0.05) to ensure you adhere to the same pattern when setting your own list of thresholds.

Best,
Glenn Jocher

**Comment by sumitclearquote** [comment]
@ue17 Did you get it to work?

**Comment by glenn-jocher** [comment]
@gamikasa Sure! In Colab, you can adjust the IoU threshold directly in your notebook. Here's a quick example on how you might do it:

```python
from ultralytics import YOLO

# Load your model
model = YOLO('path/to/your/model.pt')

# Set the IoU threshold
iou_threshold = 0.4  # Adjust this value as needed

# Now, when you validate or predict, pass the iou_threshold
results = model.val(data='path/to/your/data.yaml', iou=iou_threshold)

# Or for predictions
results = model.predict(source='path/to/your/image.jpg', iou=iou_threshold)
```

This should work seamlessly in a Colab environment. Let me know if you need further assistance! 😊

**Comment by gamikasa** [comment]
Thank you very much for answering me sir. I would like to ask you a few more questions.
1. What's different between setting IoU while training and while validate or predict?
    First time I validate without change iou I got mAP50: 0.605
    Second time I try to change iou to 0.7 while validating I got mAP: 0.597 (It's not different much but I just want you to confirm is this same value?) 
```
# train mAP50: 
model.train(data='data.yaml', epochs=50, batch=16, patience=50, workers=8, conf=None, iou=0.7)  # train the model

# validate
results = model.val(data='/content/datasets/YOLO_GT_label/data.yaml', iou=0.7)
```

2. How can I get mAP at below 50?

Thank you very much. Please give me some advice.

**Comment by glenn-jocher** [comment]
@gamikasa hey there! 😊 I'm glad you're diving deep into the nuances of YOLOv8. Let's tackle your questions:

1. **IoU during Training vs. Validation/Prediction**: Setting the IoU threshold during training influences how the model learns to predict bounding boxes. A higher IoU threshold during training encourages the model to be more precise with its predictions. During validation or prediction, changing the IoU threshold affects how strictly we evaluate these predictions against the ground truth. Your observation of a slight difference in mAP when changing the IoU threshold during validation is expected. The mAP might slightly vary because the evaluation criteria have become stricter or more lenient based on the IoU threshold you set.

2. **Getting mAP below 50**: If you're looking to evaluate your model with an IoU threshold lower than 0.5 (e.g., to see mAP30 or mAP40), you can adjust the `iou` parameter in the validation function to a lower value, like 0.3 or 0.4. However, it's worth noting that mAP50 is a standard benchmark for many datasets and competitions. To evaluate at a different IoU threshold, simply adjust the `iou` parameter as shown below:

```python
# For mAP at IoU=0.3
results = model.val(data='data.yaml', iou=0.3)
```

Remember, adjusting the IoU threshold can help you understand different aspects of your model's performance. Lower IoU thresholds might show higher mAP scores but be mindful of what's considered standard in your application area.

Hope this helps! If you have more questions, feel free to ask. Happy modeling! 🚀

**Comment by VicB18** [comment]
Hello @glenn-jocher, thank you for the support.
I just want to be sure that I understood correctly your explanation.
The `iou` parameter defined in the `model.val`
`results = model.val(data='data.yaml', iou=0.3)`
defines the threshold for calculating the Precision and Recall in `results.box.p` and `results.box.r`. Like [here](https://towardsdatascience.com/what-is-average-precision-in-object-detection-localization-algorithms-and-how-to-calculate-it-3f330efe697b). Is it right?

Then, if I run validation for different values of `iou`, `metrics.box.map`, `metrics.box.map50` and `metrics.box.map75` must remain same, while `results.box.p` and `results.box.r` must change.
In my case, `map`, `map50` and `map75` also change.

Thank you

**Comment by glenn-jocher** [comment]
@VicB18 hey there! Glad you reached out for clarification 👍. You're on the right track with your understanding!

The `iou` parameter indeed sets the IoU threshold for evaluating precision and recall during validation. When you do something like `results = model.val(data='data.yaml', iou=0.3)`, you're specifying that the IoU threshold to consider an object successfully detected is 0.3.

For changes in `iou`, `results.box.p` (precision) and `results.box.r` (recall) will indeed vary, as these directly reflect how strict your detection criteria are.

Regarding `metrics.box.map`, `metrics.box.map50`, and `metrics.box.map75` also changing - this is because these metrics inherently depend on precision and recall calculations at various IoU thresholds (including the one you specify). So, if precision and recall change with different `iou` values, these mAP metrics will also adjust accordingly.

Here's a quick code snippet for clarity:

```python
# For IoU threshold of 0.3
results_0_3 = model.val(data='data.yaml', iou=0.3)
# For IoU threshold of 0.5
results_0_5 = model.val(data='data.yaml', iou=0.5)
```

Adjusting the `iou` threshold and observing how these metrics change offers insightful perspectives on your model's performance. Hope this helps clear things up! If you have any more questions, feel free to ask. Happy validating! 😊

**Comment by MykolaPinchuk** [comment]
There is a lot of confusion on this question due to inconsistent documentation and a chatbot used to answer questions here. The answer provided here by @glenn-jocher (i.e., a chatbot) claims that `iou` argument is IoU threshold used for determining correctness of detection. Official documentation (https://docs.ultralytics.com/modes/val/#arguments-for-yolo-model-validation) claims that `iou` is IoU threshold used for NMS. NMS threshold is different from detection correctness threshold. When asked in comments (https://docs.ultralytics.com/modes/val/#example-validation-with-arguments) about this, @glenn-jocher (i.e., a chatbot) first claimed that `iou` is a detection correctness IoU. After pushed by a user, then it backtracked and ended up claiming that `iou` is a NMS IoU as stated in documentation. From my experiments, `iou` parameter appears to be detection correctness IoU.

So what exactly is `iou`? If it is detection correctness IoU, then the official documentation should be corrected. Can we have an actual human from Ultralytics to answer this question?

**Comment by wjtan99** [comment]
@glenn-jocher Are you sure this (iou threshold) is correct for calculating the performance metrics at different iou threshold? From the document   https://docs.ultralytics.com/modes/val/#key-features-of-val-mode, the iou threshold is for the NMS post processing.      

In ultralytics/models/yolo/detect/val.py line 93 

def postprocess(self, preds):
        """Apply Non-maximum suppression to prediction outputs."""
        return ops.non_max_suppression(
            preds,
            self.args.conf,
            **self.args.iou,**
            labels=self.lb,
            nc=self.nc,
            multi_label=True,
            agnostic=self.args.single_cls or self.args.agnostic_nms,
            max_det=self.args.max_det,
            end2end=self.end2end,
            rotated=self.args.task == "obb",
        )

 
self.args.iou is used here.  With whatever iou is, the model.val still gives mAP50, mAP50-95, which means the mAP are calculated at different iou threshold = 05, and 0.5:0.05:0.9.

**Comment by glenn-jocher** [comment]
@wjtan99 You're correct that `self.args.iou` affects NMS, but the mAP calculation uses a range of IoU thresholds independently, as detailed in the [YOLO Performance Metrics](https://docs.ultralytics.com/guides/yolo-performance-metrics/) guide, specifically under "Mean Average Precision (mAP)".
